Code/Tools/AssetProcessor/native/FileWatcher/FileWatcher.cpp

```diff
@@ -6,154 +6,119 @@
  *
  */
 #include "FileWatcher.h"
+#include "AzCore/std/containers/vector.h"
 #include <native/assetprocessor.h>
+#include <native/FileWatcher/FileWatcher_platform.h>
+#include <QFileInfo>
 
-//////////////////////////////////////////////////////////////////////////////
-/// FolderWatchRoot
-void FolderRootWatch::ProcessNewFileEvent(const QString& file)
+//! IsSubfolder(folderA, folderB)
+//! returns whether folderA is a subfolder of folderB
+//! assumptions: absolute paths, case insensitive
+static bool IsSubfolder(const QString& folderA, const QString& folderB)
 {
-    FileChangeInfo info;
-    info.m_action = FileAction::FileAction_Added;
-    info.m_filePath = file;
-    const bool invoked = QMetaObject::invokeMethod(m_fileWatcher, "AnyFileChange", Qt::QueuedConnection, Q_ARG(FileChangeInfo, info));
-    Q_ASSERT(invoked);
-}
+    // lets avoid allocating or messing with memory - this is a MAJOR hotspot as it is called for any file change even in the cache!
+    int sizeB = folderB.length();
+    int sizeA = folderA.length();
 
-void FolderRootWatch::ProcessDeleteFileEvent(const QString& file)
-{
-    FileChangeInfo info;
-    info.m_action = FileAction::FileAction_Removed;
-    info.m_filePath = file;
-    const bool invoked = QMetaObject::invokeMethod(m_fileWatcher, "AnyFileChange", Qt::QueuedConnection, Q_ARG(FileChangeInfo, info));
-    Q_ASSERT(invoked);
-}
+    if (sizeA <= sizeB)
+    {
+        return false;
+    }
 
-void FolderRootWatch::ProcessModifyFileEvent(const QString& file)
-{
-    FileChangeInfo info;
-    info.m_action = FileAction::FileAction_Modified;
-    info.m_filePath = file;
-    const bool invoked = QMetaObject::invokeMethod(m_fileWatcher, "AnyFileChange", Qt::QueuedConnection, Q_ARG(FileChangeInfo, info));
-    Q_ASSERT(invoked);
+    QChar slash1 = QChar('\\');
+    QChar slash2 = QChar('/');
+    int posA = 0;
+
+    // A is going to be the longer one, so use B:
+    for (int idx = 0; idx < sizeB; ++idx)
+    {
+        QChar charAtA = folderA.at(posA);
+        QChar charAtB = folderB.at(idx);
+
+        if ((charAtB == slash1) || (charAtB == slash2))
+        {
+            if ((charAtA != slash1) && (charAtA != slash2))
+            {
+                return false;
+            }
+            ++posA;
+        }
+        else
+        {
+            if (charAtA.toLower() != charAtB.toLower())
```

Comment:
This is fixed with the latest commits. It isn't using `AZ::IO::Path` because that would have to copy both strings. But it follows the decision that `AZ::IO::Path` makes and uses case-sensitive comparisons if the path separator is `/`. IMHO this is still incorrect, as that results in case-sensitive comparisons on Mac, but that decision was already made.